src/lib/dom-api-mutation.html

```diff
@@ -0,0 +1,280 @@
+      this.domApi = domApi;
+      this.node = this.domApi.node;
+      this._listeners = [];
+      this._addedNodes = [];
+      this._removedNodes = [];
+    };
+
+    DomApi.Mutation.prototype = {
+
+      addListener: function(callback, options) {
+        options = options || {};
+        this._ensureSetup(options);
+        var listener = {fn: callback, options: options};
+        this._notifyInitial(listener);
+        return this._listeners.push(listener);
+      },
+
+      removeListener: function(handle) {
+        this._listeners.splice(handle - 1, 1);
+        if (!this._hasListeners()) {
+          this._ensureCleanup();
+        }
+      },
+
+      _ensureSetup: function(options) {
+        if (!this._isSetup) {
+          this._isSetup = true;
+          this._observeContentElements(this.domApi.childNodes);
+        }
+      },
+
+      _ensureCleanup: function() {}, // abstract
+
+      _hasListeners: function() {
+        return Boolean(this._listeners.length);
+      },
+
+      addNode: function(node) {
+        if (this._hasListeners()) {
+          this._addedNodes.push(node);
+          this._scheduleNotify();
+        }
+      },
+
+      removeNode: function(node) {
+        if (this._hasListeners()) {
+          this._removedNodes.push(node);
+          this._scheduleNotify();
+        }
+      },
+
+      addAllNodes: function(force) {
+        if (this._hasListeners() || force) {
+          var c$ = this.domApi.childNodes;
+          if (c$.length) {
+            for (var i=0, c; (i < c$.length) && (c=c$[i]); i++) {
+              this._addedNodes.push(c);
+            }
+            this._scheduleNotify();
+          }
+        }
+      },
+
+      _scheduleNotify: function() {
+        this._debouncer = Polymer.Debounce(this._debouncer, 
+          this._notify);
+        this._debouncer.context = this;
+        Polymer.dom.addDebouncer(this._debouncer);
+      },
+
+      _notify: function(mxns) {
+        var info = mxns || {
+          target: this.node,
+          addedNodes: this._addedNodes,
+          removedNodes: this._removedNodes
+        }
+        this._updateContentElements(info);
+        this._callListeners(info);
+        this._addedNodes = [];
+        this._removedNodes = [];
+      },
+
+      _notifyInitial: function(listener) {
+        var info = {
+          target: this.node,
+          addedNodes: this.domApi.childNodes,
+          removedNodes: []
+        };
+        this._callListener(listener, info);
+      },
+
+      _updateContentElements: function(info) {
+        this._observeContentElements(info.addedNodes);
+        this._unobserveContentElements(info.removedNodes);
+      },
+
+      _observeContentElements: function(elements) {
+        for (var i=0, h, n; (i < elements.length) && (n=elements[i]); i++) {
+          if (this._isContent(n)) {
+            n.__observeNodesMap = n.__observeNodesMap || new WeakMap();
+            if (n.__observeNodesMap.get(this) == null) {

```

Comment:
Is the check necessary? If everything is working correctly, should only ever observe one content once?